### PR TITLE
xtensa/esp32: Add AES hardware accelerator driver

### PR DESCRIPTION
**Pull request by donghengqaz**
## Summary

Add AES hardware accelerator driver.

## Testing

boards/xtensa/esp32/esp32-devkitc/configs/aes/defconfig.

### CODE REVIEW

**Comment by Ouss4** [comment]
@donghengqaz the defconfig supplied needs to be refreshed: `./tools/refresh.sh --silent esp32-devkitc/aes`

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 1]
```diff
@@ -0,0 +1,1051 @@
...
+  putreg32(buffer[0], AES_TEXT_BASE + 0);
+  putreg32(buffer[1], AES_TEXT_BASE + 4);
+  putreg32(buffer[2], AES_TEXT_BASE + 8);
+  putreg32(buffer[3], AES_TEXT_BASE + 12);
+
+  putreg32(AES_START_OPT, AES_START_REG);
+
+  while (getreg32(AES_IDLE_REG) != AES_IDLE_STATE)
+    {
+      ;
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:117]: This ";" inside is not necessary. You can put it after the "}" if you want, but also not necessary!

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 2]
```diff
@@ -0,0 +1,1051 @@
...
+
+      ret = esp32_aes_ecb_cypher(&aes, 0, encrypt_buf, decrypt_buf, size);
+      assert(ret == 0);
+
+      if (memcmp(decrypt_buf, input, size))
+        {
+          assert(0);
+        }
+
+      printf("ESP32 AES ECB key=%d bits test: PASS\n", keybits);
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:725]: Don't use printf() in the kernel

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 3]
```diff
@@ -0,0 +1,1051 @@
...
+      ret = esp32_aes_cbc_cypher(&aes, 0, iv_buf, encrypt_buf,
+                                 decrypt_buf, size);
+      assert(ret == 0);
+
+      if (memcmp(decrypt_buf, input, size))
+        {
+          assert(0);
+        }
+
+      printf("ESP32 AES CBC key=%d bits test: PASS\n", keybits);
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:811]: Don't use printf() in the kernel

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 4]
```diff
@@ -0,0 +1,1051 @@
...
+      ret = esp32_aes_ctr_cypher(&aes, &nc_off, cnt_buf, cache_buf,
+                                 encrypt_buf, decrypt_buf, size);
+      assert(ret == 0);
+
+      if (memcmp(decrypt_buf, input, size))
+        {
+          assert(0);
+        }
+
+      printf("ESP32 AES CTR key=%d bits test: PASS\n", keybits);
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:902]: Don't use printf() in the kernel

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 5]
```diff
@@ -0,0 +1,1051 @@
...
+      ret = esp32_aes_xts_cypher(&aes, false, unit_buf, encrypt_buf,
+                                 decrypt_buf, size);
+      assert(ret == 0);
+
+      if (memcmp(decrypt_buf, input, size))
+        {
+          assert(0);
+        }
+
+      printf("ESP32 AES XTS key=%d bits test: PASS\n", keybits);
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:1029]: Ditto

File: arch/xtensa/src/esp32/esp32_aes.c [hunk 6]
```diff
@@ -0,0 +1,1051 @@
...
+int esp32_aes_main(int argc, char *argv[])
+{
+  esp32_aes_init();
+
+  esp32_aes_ecb_test();
+  esp32_aes_cbc_test();
+  esp32_aes_ctr_test();
+  esp32_aes_xts_test();
+
+  printf("\nESP32 AES hardware accelerate test done.\n");
```
Comment (acassis) [inline arch/xtensa/src/esp32/esp32_aes.c:1046]: Ditto

**Review by acassis — CHANGES_REQUESTED** [review]

**Pull request merged**